src/Microsoft.VisualStudio.Services.Agent/Util/ServerUtil.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.WebApi;
+
+namespace Microsoft.VisualStudio.Services.Agent.Util
+{
+    public static class ServerUtil
+    {
+        private static DeploymentFlags _deploymentType;
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA2000:Dispose objects before losing scope", MessageId = "locationServer")]
+        private static async Task<Location.ConnectionData> GetConnectionData(string serverUrl, VssCredentials credentials, ILocationServer locationServer)
+        {
+            VssConnection connection = VssUtil.CreateConnection(new Uri(serverUrl), credentials);
+            await locationServer.ConnectAsync(connection);
+            return await locationServer.GetConnectionDataAsync();
+        }
+
+        public static bool IsHosted()
+        {
+            // If deployment type was not determined before, an exception will be thrown
+            switch (_deploymentType)
+            {
+                case DeploymentFlags.Hosted:
+                    return true;
+                case DeploymentFlags.OnPremises:
+                    return false;
+                case DeploymentFlags.None:
+                    throw new Exception($"Deployment type has not been determined");
```

Comment:
Is it expected that previously IsHostedServer didn't throw an exception, but currently it can throw?

Reply:
Currently, the `IsHosted` method is designed to be called with arguments to define the server deployment type or to be called without them, if the type has already been defined. Once the type is determined, it becomes a constant for the current process, and the method just returns this constant, no matter it is called with or without arguments. This solution is useful, e.g. in loops, where the method with arguments has to be called for several times — since the operation will be time-consuming only on the first iteration. The method without arguments worths being called only if the type has already been defined, otherwise there will be an exception thrown.

Also, an exception will be thrown, if there will be another server deployment type determined, different from `Hosted` and `OnPremises`. In this case, it is not a good idea just to say that the server is hosted, when it is not.

Reply:
@DenisRumyantsev keeping state as static is not a good practice (since this is not flexible solution for future changes usually) - could we avoid this?
It's also not obvious that isHosted() signature keeps previous deployment type value.

Reply:
I reasoned this way during development:

We have a loop with retries.
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Agent.Listener/SelfUpdater.cs#L233

Inside this loop, the `HashValidation` method is called.
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Agent.Listener/SelfUpdater.cs#L299

Inside the `HashValidation` method, the `IsHosted` method with arguments is called.
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Agent.Listener/SelfUpdater.cs#L177

Since it is enough to determine the server type for the executed process once (at the first iteration), save it, and simply return this value at subsequent iterations, instead of calculating the same thing each time with an already calculated result, I decided to implement the `IsHosted` method with arguments exactly in this way.
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Microsoft.VisualStudio.Services.Agent/Util/ServerUtil.cs#L39

I did not implement the `IsHosted` method as flexible, since the server type cannot change during running the agent.

I also decided to leave the `IsHosted` method without arguments as public, because I found an already existing similar logic.
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Microsoft.VisualStudio.Services.Agent/Util/ServerUtil.cs#L23

For example, in the public `GetConnectionDataAsync` method, it is customary to throw an exception if another `ConnectAsync` method has not been called before.
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Microsoft.VisualStudio.Services.Agent/LocationServer.cs#L55
https://github.com/microsoft/azure-pipelines-agent/blob/0e87b9cd1c639b4c4342ce2ac012f8d5562c5d8c/src/Microsoft.VisualStudio.Services.Agent/LocationServer.cs#L27

Reply:
@DenisRumyantsev yes, throwing an exception and making it public looks fine since we described it in summary comments, although the method you mentioned is not static, and relevant class does not keep the state in static fields.

Reply:
Fixed